File: soln-stack-problems/validate-stack-sequences.java
/*
https://leetcode.com/problems/validate-stack-sequences/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Stack<Integer> stack=new Stack<>();        
        int index=0;
        
        for(int i=0;i<pushed.length;i++){
            stack.push(pushed[i]);
            while(!stack.isEmpty() && popped[index]==stack.peek() && index<popped.length){
                stack.pop();
                index++;
            }                
        }
        return stack.isEmpty()? true:false;        
    }
}

File: C++/soln-stack-problems/validate-stack-sequences.cpp
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/validate-stack-sequences/
*/

class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> stack;     
        int index=0;
        
        for(int i=0;i<pushed.size();i++){
            stack.push(pushed[i]);
            while(!stack.empty() && popped[index]==stack.top() && index<popped.size()){
                stack.pop();
                index++;
            }                
        }
        return stack.empty()? true:false;        
    }
};